pkg/controllers/inventory/inventory.go

```diff
@@ -0,0 +1,124 @@
+	}
+	inventory.vpcCache = cache.NewIndexer(
+		func(obj interface{}) (string, error) {
+			vpc := obj.(*runtimev1alpha1.Vpc)
+			return fmt.Sprintf("%v/%v-%v", vpc.Namespace, vpc.Labels[VpcLabelAccountName], vpc.Info.Id), nil
+		},
+		cache.Indexers{
+			VpcIndexerByAccountNameSpacedName: func(obj interface{}) ([]string, error) {
+				vpc := obj.(*runtimev1alpha1.Vpc)
+				return []string{vpc.Namespace + "/" + vpc.Labels[VpcLabelAccountName]}, nil
+			},
+			VpcIndexerByVpcNamespacedName: func(obj interface{}) ([]string, error) {
+				vpc := obj.(*runtimev1alpha1.Vpc)
+				return []string{vpc.Namespace + "/" + vpc.Name}, nil
+			},
+			VpcIndexerByNamespacedRegion: func(obj interface{}) ([]string, error) {
+				vpc := obj.(*runtimev1alpha1.Vpc)
+				return []string{vpc.Namespace + "/" + vpc.Info.Region}, nil
+			},
+			VpcIndexerByNamespace: func(obj interface{}) ([]string, error) {
+				vpc := obj.(*runtimev1alpha1.Vpc)
+				return []string{vpc.Namespace}, nil
+			},
+		})
+	return inventory
+}
+
+// BuildVpcCache using vpc list from cloud, update vpc cache(with vpcs applicable for the current cloud account).
+func (inventory *Inventory) BuildVpcCache(vpcMap map[string]*runtimev1alpha1.Vpc, namespacedName *types.NamespacedName) error {
+	vpcsInCache, err := inventory.vpcCache.ByIndex(VpcIndexerByAccountNameSpacedName, namespacedName.String())
+	if err != nil {
+		return fmt.Errorf("fetching vpc list from vpc cache failed, "+
+			"index %v, error %v", *namespacedName, err)
+	}
+
+	// Remove vpcs in vpc cache which are not found in vpc list fetched from cloud.
+	for _, i := range vpcsInCache {
+		vpc := i.(*runtimev1alpha1.Vpc)
+		if _, found := vpcMap[vpc.Info.Id]; !found {
+			inventory.log.V(1).Info("Deleting a vpc from vpc cache", "vpc id", vpc.Info.Id, "account",
+				namespacedName.String())
+			if err := inventory.vpcCache.Delete(vpc); err != nil {
+				inventory.log.Error(err, "failed to delete entry from VpcIndexer", "vpc id", vpc.Info.Id, "account",
+					namespacedName.String())
+			}
+		}
+	}
+
+	for _, v := range vpcMap {
+		err := inventory.vpcCache.Add(v)
+		if err != nil {
+			return fmt.Errorf("failed to add entry into VpcIndexer, vpc id %s, account %v, error %v",
+				v.Info.Id, *namespacedName, err)
+		}
+	}
+
+	return nil
+}
+
+// DeleteVpcCache deletes all entries from vpc cache.
+func (inventory *Inventory) DeleteVpcCache(namespacedName *types.NamespacedName) error {
+	vpcsInCache, err := inventory.vpcCache.ByIndex(VpcIndexerByAccountNameSpacedName, namespacedName.String())
+	if err != nil {
+		return err
+	}
+	for _, i := range vpcsInCache {
+		vpc := i.(*runtimev1alpha1.Vpc)
+		if err := inventory.vpcCache.Delete(vpc); err != nil {
+			return fmt.Errorf("failed to delete entry from VpcIndexer, indexer %v, error %v", *namespacedName, err)
+		}
+	}
+	return nil
+}
+
+// GetVpcCacheIndexer returns the vpc cache indexer.
+func (inventory *Inventory) GetVpcCacheIndexer() cache.Indexer {
```

Comment:
I am actually not comfortable with returning indexer outside this package. Its better to have get calls from api server and we do processing here inside this package

Reply:
Added apis in inventory.go for processing request from rest, hence avoiding passing indexer.